syntax = "proto3";

package customers;

service Customers{

rpc UpdateSellerFeedback(SellerFeedbackMessage) returns (generalResponse){}
rpc GetUserDB(UserCredentialsMessage) returns (generalResponse){}
rpc RegisterSellerDB(UserCredentialsMessage) returns (generalResponse){}

rpc CreateAccount(CreateAccountRequestMessage) returns (generalResponse){}
rpc Login(LoginRequestMessage) returns (generalResponse){}
rpc AddToCart(AddToCartRequestMessage) returns (generalResponse){}
rpc RemoveItemFromCart(RemoveFromCartRequestMessage) returns (generalResponse){}
rpc ClearCart(ClearCartRequestMessage) returns (generalResponse){}
rpc DisplayCart(DisplayCartRequestMessage) returns (DisplayCartResponseMessage){}
rpc MakePurchase(MakePurchaseMessage) returns (generalResponse){}
rpc PurchaseHistory(PurchaseHistoryMessage) returns (generalResponse){}
rpc ProvideFeedback(ProvideFeedbackMessage) returns (generalResponse){}
rpc GetSellerRating(GetSellerRatingMessage) returns (generalResponse){}
rpc Logout(LogoutMessage) returns (generalResponse){}
rpc GetBuyerId(GetBuyerIdRequestMessage) returns (GetBuyerIdResponseMessage){}
rpc SetLoginState(SetLoginStateRequestMessage) returns (generalResponse){}
rpc HasProvidedFeedback(HasProvidedFeedbackRequestMessage) returns(HasProvidedFeedbackResponseMessage){}
rpc UpdateCustomerProvideFeedback(UpdateProvideFeedbackRequestMessage) returns (generalResponse){}
rpc UpdateSellerFeedbackFromBuyer(UpdateSellerFeedbackRequestMessage) returns (generalResponse){}

}


message generalResponse{
    string msg=1;
}

message CreateAccountRequestMessage{
  string username = 1;
  string password = 2;
  string name = 3;
}

message LoginRequestMessage{
  string username = 1;
  string password = 2;
}

message SearchProductMessage{
    int32 item_category = 1;
    string keywords = 2;
}

message AddToCartRequestMessage{
  int32 buyer_id = 1;
  int32 product_id = 2;
  int32 quantity = 3;
}

message RemoveFromCartRequestMessage{
  int32 buyer_id = 1;
  int32 product_id = 2;
  int32 quantity = 3;
}

message ClearCartRequestMessage{
    int32 buyer_id = 1;
}

message DisplayCartRequestMessage{
    int32 buyer_id = 1;
}

message MakePurchaseMessage{
    int32 buyer_id = 1;
    string credit_card = 2;
}

message PurchaseHistoryMessage{
    int32 buyer_id = 1;
}

message ProvideFeedbackMessage{
    int32 buyer_id =1;
    int32 product_id = 2;
    string feedback_type = 3;
}

message GetSellerRatingMessage{
    int32 seller_id = 1;
}

message LogoutMessage{
    int32 buyer_id = 1;
}

message GetBuyerIdRequestMessage{
    string username = 1;
}

message GetBuyerIdResponseMessage{
    int32 buyer_id = 1;
}

message SetLoginStateRequestMessage{
  int32 buyer_id = 1; 
  bool state = 2;
}

message CartItem {
  int32 product_id = 1;
  int32 quantity = 8;
}

message DisplayCartResponseMessage {
  repeated CartItem cart_items = 1;
}

message HasProvidedFeedbackRequestMessage{
  int32 buyer_id = 1;
  int32 product_id = 2;
}

message HasProvidedFeedbackResponseMessage {
    bool has_provided = 1;
}

message UpdateProvideFeedbackRequestMessage{
  int32 buyer_id = 1;
  int32 product_id = 2;
}

message UpdateSellerFeedbackRequestMessage{
  int32 seller_id = 1;
  int32 feedback_type = 2;
}

// *********************************
// SELLER: Assignment 4 updates
//**********************************
message UserCredentialsMessage{
  string username = 1;
  optional string password = 2;
}

message SellerFeedbackMessage{
  int32 seller_id = 1;
  int32 tu = 2;
  int32 td = 3;
}
